javaslang/src/main/java/javaslang/collection/TreeMap.java

```diff
@@ -477,6 +505,98 @@ public boolean isEmpty() {
         return TreeSet.ofAll(keyComparator(), iterator().map(Tuple2::_1));
     }
 
+    @Override
+    public Iterator<Tuple2<K, V>> descendingIterator() {
+        return entries.descendingIterator();
+    }
+
+    @Override
+    public SortedSet<K> descendingKeySet() {
+        return descendingMap().keySet();
+    }
+
+    @Override
+    public SortedMap<K, V> descendingMap() {
+        return new TreeMap<>(entries.descendingTree());
+    }
+
+    @Override
+    public SortedMap<K, V> subMap(Option<K> fromKey, boolean fromInclusive, Option<K> toKey, boolean toInclusive) {
+        return new TreeMap<>(
+                entries.subTree(
+                    fromKey.map(k -> Tuple.of(k, null)), fromInclusive,
+                    toKey.map(k -> Tuple.of(k, null)), toInclusive));
+    }
+
+    @Override
+    public SortedMap<K, V> subMap(K fromKey, boolean fromInclusive, K toKey, boolean toInclusive) {
+        return subMap(Option.some(fromKey), fromInclusive, Option.some(toKey), toInclusive);
+    }
+
+    @Override
+    public SortedMap<K, V> subMap(K fromKey, K toKey) {
+        return subMap(fromKey, true, toKey, false);
+    }
+
+    @Override
+    public SortedMap<K, V> headMap(K toKey, boolean inclusive) {
+        return subMap(Option.none(), true, Option.some(toKey), inclusive);
+    }
+
+    @Override
+    public SortedMap<K, V> headMap(K toKey) {
+        return headMap(toKey, false);
+    }
+
+    @Override
+    public SortedMap<K, V> tailMap(K fromKey, boolean inclusive) {
+        return subMap(Option.some(fromKey), inclusive, Option.none(), true);
+    }
+
+    @Override
+    public SortedMap<K, V> tailMap(K fromKey) {
+        return tailMap(fromKey, true);
```

Comment:
This is a hard design decision. I understand you both. Nobody likes lists of boolean params, they are meaningless. However, Javaslang is still a Java library and devs are used to the existing NavigableMap API.

Adding an `enum Inclusion { INCLUSIVE, EXCLUSIVE }` would solve the problem but increase the number of types. Where should it be placed? I should not be a top-level type. The main reason Javaslang looks relatively simple and compact is because helper classes/types are hidden as package-private top-level classes or public static nested classes. I think the reason why the Java language architects used boolean may also be the fact that there is no clean place to put such a type.

In order to keep it simple I would follow the Java interface, as @eduardmanas suggested.